### PR TITLE
feat(core): Introduce TestBed.inject to replace TestBed.get

### LINKED ISSUE #26491: Use typed get method in TestBed class
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

The `get` static method of `TestBed` class does not return a typed object.

```ts
const service = TestBed.get(MyService); // service is typed as any
```

## Expected behavior

The `get` static method should return a typed object.

```ts
const service = TestBed.get(MyService); // service should be typed as MyService
```

## Minimal reproduction of the problem with instructions

Look examples.

## What is the motivation / use case for changing the behavior?

This would improve type safety.

## Environment

<pre><code>
Angular version: 6.1.9

</code></pre>

## Others:

A possible implementation using new helper types from TypeScript would be:

```ts
    static get<T extends new (...args: any[]) => any>(token: T, notFoundValue?: T | InstanceType<T>): InstanceType<T>;
```

I'm not sure if the `notFoundValue` should be a constructor or an instance, so I typed it as both, but fell free to update it as is actually needed. I tested that signature and it actually returns a typed object as wanted.

### LINKED ISSUE #29905: New TestBed.get shows deprecation warning for abstract types and everywhere if using a strict tsconfig
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core/testing

### Is this a regression?

Yes, the previous version in which this bug was not present was: `8.0.0-beta.11`

### Description

`8.0.0-beta.12` introduced a new `TestBed.get` method and deprecated the existing one (see 
https://github.com/angular/angular/commit/609024f93dc2df038ca5c3d8bfc2406ca5a2a6ac):

https://github.com/angular/angular/blob/609024f93dc2df038ca5c3d8bfc2406ca5a2a6ac/packages/core/testing/src/test_bed.ts#L246-L256

The new signature should be transparent for users as `const service = TestBed.get(UserService);` works.

But the new signature does not work with abstract types, like for example `HttpTestingController` (see https://github.com/angular/angular/blob/master/packages/common/http/testing/src/api.ts#L29) which leads to confusing deprecation warnings.

Update: this is even happening on every `TestBed.get` calls if you have a strict tsconfig, see https://github.com/angular/angular/issues/29905#issuecomment-483651602

## 🔬 Minimal Reproduction

```typescript
import { TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';

describe('UserService', () => {
  beforeEach(() => TestBed.configureTestingModule({
    imports: [HttpClientTestingModule]
  }));

  it('should be created', () => {
    const http: HttpTestingController = TestBed.get(HttpTestingController);
    expect(http).toBeTruthy();
  });
});
```

## 🔥 Exception or Error

`ng lint` shows:

```
/src/app/user.service.spec.ts:10:49 - get is deprecated: from v8.0.0 use Type<T> or InjectionToken<T>
```

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 8.0.0-beta.13
Node: 11.11.0
OS: darwin x64
Angular: 8.0.0-beta.12
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.0-beta.13
@angular-devkit/build-angular     0.800.0-beta.13
@angular-devkit/build-optimizer   0.800.0-beta.13
@angular-devkit/build-webpack     0.800.0-beta.13
@angular-devkit/core              8.0.0-beta.13
@angular-devkit/schematics        8.0.0-beta.13
@angular/cli                      8.0.0-beta.13
@ngtools/webpack                  8.0.0-beta.13
@schematics/angular               8.0.0-beta.13
@schematics/update                0.800.0-beta.13
rxjs                              6.4.0
typescript                        3.4.3
webpack                           4.29.6
</code></pre>

cc @Goodwine who wrote it and @mhevery who reviewed: I have not followed closely but couldn't `get<T>(token: T | InjectionToken<T>, notFoundValue?: T, flags?: InjectFlags): any;` be enough?

### PR DESCRIPTION
**Pull request by Goodwine**
TestBed.get is not type safe, fixing it would be a massive breaking
change. The Angular team has proposed replacing it with TestBed.inject
and deprecate TestBed.get.

Fixes #26491
Fixes #29905

BREAKING CHANGE: Injector.get now accepts abstract classes to return
type-safe values. Previous implementation returned `any` through the
deprecated implementation.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

```ts
TestBed.get(ChangeDetectorRef) // returns any
```

Issue Number: #26491, #29905

## What is the new behavior?
```ts
// Now deprecated
TestBed.get(ChangeDetectorRef) // returns any

TestBed.inject(ChangeDetectorRef) // returns ChangeDetectorRef
```

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

The breaking change affects a feature that has been deprecated since Angular 4.x
`Injector#get()` has 2 signatures:
- Typed signature, accept an `InjectionToken<T>` and a class (`Type<T>`)
- Anything else is also accepted but typed as `any`

The updated Typed signature now also accepts abstract classes (`AbstractType<T>`) which may cause compilation issues to the deprecated use of `any` as it would need to be manually downcasted to the concrete type if needed.

## Other information

Internal doc - [go/testbed.inject](http://go/testbed.inject)

**Note:** I did not update _all_ uses of `TestBed.get` in the Angular codebase to keep this diff small.
I will follow with another CL migrating all `TestBed.get` to `TestBed.inject` if this PR goes through.

### CODE REVIEW

File: aio/src/app/app.component.spec.ts [hunk 1]
```diff
@@ -677,7 +677,7 @@ describe('AppComponent', () => {
       });
 
       it('should not be loaded/registered until necessary', () => {
-        const loader: TestElementsLoader = fixture.debugElement.injector.get(ElementsLoader);
+        const loader = fixture.debugElement.injector.get(ElementsLoader) as unknown as TestElementsLoader;
```
Comment (mhevery) [inline aio/src/app/app.component.spec.ts]: I don't understand why this change is needed.

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:264658383:BASE:264658421:1566413356611:5859f5f1)

File: aio/src/app/app.component.spec.ts (hunk 1 above)
Comment (Goodwine) [reply aio/src/app/app.component.spec.ts]: Per the "breaking changes" comment, I added `AbstractType<T>` to the "typed version" of `Injector#get`. Previously abstract classes would have returned `any`, the untyped version has been deprecated since Angular 4.x

This caused the returned value here to be type `ElementsLoader` which I thought I could cast directly to `TestElementsLoader` but they are apparently not fully compatible since the compiler complained about it

**Comment by mhevery** [comment]
This PR breaks a lot of things in `google3` need to investigate more before it can be merged.

**Comment by Goodwine** [comment]
It can be caused by:
- Change in `Injector#get`, abstract classes now return the abstract type and need to be downcasted, this could cause compilation errors, I looked at the breakages and it seems like it's unproper typing that caused it.
- It will make all `TestBed.get` calls deprecated which would "break" all presubmits that block on `TSLint`

The former can be fixed by using proper typings, the latter can be solved by not marking `TestBed.get` in this commit and do it after migrating all of google3

**Comment by Goodwine** [comment]
Discussed with Misko, I'll remove the deprecation in this PR and fix breakages. Once it's merged, then I'll migrate google3 code, and then send another PR deprecating TestBed.get

File: packages/core/test/render3/view_container_ref_spec.ts [hunk 2]
```diff
@@ -295,7 +295,7 @@ describe('ViewContainerRef', () => {
              const changeDetector = ref.injector.get(ChangeDetectorRef);
              changeDetector.detectChanges();
              expect(dynamicComp.doCheckCount).toEqual(1);
-             expect(changeDetector.context).toEqual(dynamicComp);
+             expect((changeDetector as any).context).toEqual(dynamicComp);
```
Comment (IgorMinar) [inline packages/core/test/render3/view_container_ref_spec.ts]: this change tells me that this is a painful breaking change that will likely affect lots of code.

**Review by IgorMinar — COMMENTED** [review]
Can't we change this so that #get behavior is unmodified while #inject projects a type-safe implementation? 

I don't think that we can change the behavior of the #get api without causing lots of breakages.

File: packages/core/test/render3/view_container_ref_spec.ts (hunk 2 above)
Comment (Goodwine) [reply packages/core/test/render3/view_container_ref_spec.ts]: Note that this is `Injector.get`, not `TestBed`, the change here is that we're now taking abstract classes in consideration to return the right type instead of returning `any` which is the deprecated behavior since Angular 4.x

I actually don't expect many breakages from `Injector.get`

**Comment by Goodwine** [comment]
> Can't we change this so that #get behavior is unmodified while #inject projects a type-safe implementation?

This is what this commit is doing, I discussed with Vikram (see doc linked at the bottom of the PR) and he had the concern that `TestBed.get` is just too big to break, so we could add `TestBed.inject` and deprecate `TestBed.get`, the deprecation would happen on a separate commit.

File: packages/core/test/render3/view_container_ref_spec.ts (hunk 2 above)
Comment (Goodwine) [reply packages/core/test/render3/view_container_ref_spec.ts]: For example, this only broke 5~6 targets on google3, affecting about 10~15 files which I already fixed.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:265113719:BASE:265114002:1566589515684:7c385a71)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:265113719:BASE:266000752:1567029492437:7c4a57ef)

**Comment by Goodwine** [comment]
I have written the PR for that - https://github.com/angular/angular/pull/32382
One of the AIO tests was failing because the type wasn't found, I'm suspecting it needs this one submitted because I believe it might be pulling the types from HEAD, no?
